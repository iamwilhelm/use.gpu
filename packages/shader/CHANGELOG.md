0.6.0
- Add simple type inference via `@infer`
- Add struct types for storage
- Remove `/types` export

0.5.0
- Add `vec3to4` polyfill for storage
- Add `u8`/`u16`/`i8`/`i16` polyfill for storage

0.4.0
- Add WGSL support at `@use-gpu/shader/wgsl`

0.3.0
- Add `bindBundle` / `bindModule` methods to produce virtually linked bundles that acts as closures.
- Add `bindingToModule` and `bindingsToLinks` methods to convert use-gpu bindings to shaders
- Add `castTo` operation to quickly swizzle data on the inside.
- Add GLSL generation for data bindings (constant, storage, lambda).
- Add virtual and static modules, rendered just-in-time.
