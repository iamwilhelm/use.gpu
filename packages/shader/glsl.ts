export {
  parseShader,
  loadModule,
  loadModuleWithCache,
  defineConstants,
  makeLanguage,
  makeModuleCache,
} from './src/glsl/shader';

export {
  linkBundle,
  linkModule,
  linkCode,
} from './src/glsl/link';

export {
  makeASTParser,
  compressAST,
  decompressAST,
  rewriteUsingAST,
} from './src/glsl/ast';
