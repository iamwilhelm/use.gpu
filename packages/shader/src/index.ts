import GLSL from './glsl';
import WGSL from './wgsl';

export const GLSLLinker = GLSL;
export const WGSLLinker = WGSL;

export * from './util';
