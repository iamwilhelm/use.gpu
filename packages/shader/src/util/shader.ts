import { VirtualTable } from '../types';
import { getProgramHash, makeKey } from './hash';
import { PREFIX_VIRTUAL } from '../constants';

const EMPTY_LIST = [] as any[];
const EMPTY_TABLE = {} as any;

type SymbolTable = {
  symbols?: string[],
};

// Load a static (inert) module
export const loadStaticModule = (code: string, name: string, entry?: string) => {
  const table = {
    hash: getProgramHash(code),
  };
  return { name, code, table, entry };
}

// Load a virtual (generated) module
export const loadVirtualModule = <T extends SymbolTable = any>(
  virtual: VirtualTable,
  initTable: Partial<T> = EMPTY_TABLE,
  entry?: string,
  key: string | number = makeKey(),
) => {
  let symbols = initTable.symbols ?? EMPTY_LIST;
  const code = `#virtual [${symbols.join(' ')}] ${key.toString(16)}`;

  const hash = getProgramHash(code);
  const name = `${PREFIX_VIRTUAL}${hash.slice(0, 6)}_`;

  const table = {
    hash,
    symbols,
    visibles: symbols,
    ...initTable,
  };
  return { name, code, table, entry, virtual };
}