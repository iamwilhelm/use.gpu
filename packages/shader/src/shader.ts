import { Tree, SyntaxNode } from '@lezer/common';
import { parser } from './grammar/glsl';

export const parseGLSL = (code: string): Tree => {
  const parsed = parser.parse(code);
  return parsed;
}
