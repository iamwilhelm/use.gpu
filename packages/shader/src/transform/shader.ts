import { ShaderLanguages, ShaderCompiler } from '@use-gpu/core/types';
import { ParsedModule, ParsedModuleCache, ShaderDefine } from '../types';
import { Tree, SyntaxNode } from '@lezer/common';

import { makeASTParser, compressAST, decompressAST, getProgramHash } from './ast';
import { parser } from '../grammar/glsl';
import LRU from 'lru-cache';

type LangDef = {
  glsl?: ShaderCompiler,
  cache?: any,
};

// Parse a code module into its in-memory representation
// (AST + symbol table)
export const loadModule = (
  code: string,
  name: string,
  compressed: boolean = false,
): ParsedModule => {
  if (code == null) throw new Error(`Shader code ${name} undefined`);
  let tree = parseGLSL(code);

  const parser = makeASTParser(code, tree);
  const table = parser.getSymbolTable();
  const shake = parser.getShakeTable(table);

  if (compressed) tree = decompressAST(compressAST(tree));

  return {name, code, tree, table, shake};
}

// Use cache to load modules
export const loadModuleWithCache = (
  code: string,
  name: string,
  cache: ParsedModuleCache | null = null,
): ParsedModule => {
  if (!cache) return loadModule(code, name, true);

  const hash = getProgramHash(code);
  const entry = cache.get(hash);
  if (entry) return entry;
  
  const module = loadModule(code, name, true);
  cache.set(hash, module);
  return module;
}

// Parse GLSL using lezer grammar
export const parseGLSL = (code: string): Tree => parser.parse(code);

// Make GLSL definitions
export const defineGLSL = (defs: Record<string, ShaderDefine>): string => {
  const out = [];
  for (let k in defs) if (defs[k] !== false && defs[k] !== null) out.push(`#define ${k} ${defs[k]}`);
  return out.join("\n");
}

// Make shader languages interface
export const makeShaderLanguages = (langs: LangDef[]): ShaderLanguages => {
  const out = {} as any;
  for (const {glsl} of langs) {
    if (glsl) {
      out.glsl = {
        compile: (code: string, stage: any) => (glsl as any)(code, stage, false),
        cache: makeModuleCache(),
      };
    }
  }
  return out;
};

// LRU cache for parsed shader code
export const makeModuleCache = (options: Record<string, any> = {}) => new LRU({
  max: 100,
  ...options,
});
