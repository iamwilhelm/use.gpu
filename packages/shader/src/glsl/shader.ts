import { Tree } from '@lezer/common';
import { ParsedModule, ParsedModuleCache, ShaderDefine } from './types';

import { makeLoadModule, makeLoadModuleWithCache } from '../util/shader';
import { makeBundleToAttribute, makeBundleToAttributes } from '../util/bundle';
import { decompressAST } from '../util/tree';

import { makeASTParser, compressAST } from './ast';
import { toTypeString, toTypeArgs } from './type';
import { parser } from './grammar/glsl';
import LRU from 'lru-cache';

export { loadStaticModule, loadVirtualModule } from '../util/shader';

// LRU cache for parsed shader code
export const makeModuleCache = (options: Record<string, any> = {}) => new LRU<string, ParsedModule>({
  max: 100,
  ...options,
});

export const DEFAULT_CACHE = makeModuleCache();

// Parse GLSL code into lezer tree
export const parseShader = (code: string): Tree => parser.parse(code);

// Parse a code module into its in-memory representation
// (AST + symbol table)
export const loadModule = makeLoadModule(parseShader, makeASTParser, compressAST);

// Use cache to load modules
export const loadModuleWithCache = makeLoadModuleWithCache(loadModule, DEFAULT_CACHE);

// Make GLSL definitions
export const defineConstants = (defs: Record<string, ShaderDefine>): string => {
  const out = [];
  for (let k in defs) if (defs[k] !== false && defs[k] !== null) out.push(`#define ${k} ${defs[k]}`);
  return out.join("\n");
}

export const bundleToAttribute = makeBundleToAttribute(toTypeString, toTypeArgs);
export const bundleToAttributes = makeBundleToAttributes(toTypeString, toTypeArgs);
