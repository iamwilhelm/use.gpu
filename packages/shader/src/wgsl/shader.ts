import { Tree } from '@lezer/common';
import { ParsedModule, ShaderDefine } from './types';

import { makeLoadModule, makeLoadModuleWithCache } from '../util/shader';
import { makeBundleToAttribute, makeBundleToAttributes } from '../util/bundle';
import { makeTranspile } from '../util/transpile';

import { makeASTParser, compressAST, decompressAST } from './ast';
import { toTypeSymbol, toTypeArgs } from './type';
import { removeComments, removeWhiteSpace, renameLocals } from './minify';
import { parser } from './grammar/wgsl';

import LRU from 'lru-cache';
import zip from 'lodash/zip';

export { loadStaticModule, loadVirtualModule, bindEntryPoint } from '../util/shader';

/** LRU cache for parsed shader code */
export const makeModuleCache = (options: Record<string, any> = {}) => new LRU<number, ParsedModule>({
  max: 100,
  ...options,
});

export const DEFAULT_CACHE = makeModuleCache();

/** Parse WGSL code into lezer tree */
export const parseShader = (code: string): Tree => parser.parse(code);

/** Parse a code module into its in-memory representation (AST + symbol table) */
export const loadModule = makeLoadModule(parseShader, makeASTParser, compressAST, decompressAST);

/** Use cache to load modules */
export const loadModuleWithCache = makeLoadModuleWithCache(loadModule, DEFAULT_CACHE);

/** Make WGSL constant definitions */
export const defineConstants = (defs: Record<string, ShaderDefine>): string => {
  const out = [];
  for (const k in defs) if (k[0] !== '@' && defs[k] != null) out.push(`const ${k} = ${defs[k]};`);
  return out.join("\n");
}

/** Make WGSL constant definitions */
export const defineEnables = (enabled: string[]) => enabled.length ? `enable ${enabled.join(', ')};` : '';

/** Convert a bundle with a defined entry point to a definition for that attribute or type. */
export const bundleToAttribute = makeBundleToAttribute(toTypeSymbol, toTypeArgs);

/** Convert a bundle to a definition for all its attributes. */
export const bundleToAttributes = makeBundleToAttributes(toTypeSymbol, toTypeArgs);

// Simple whitespace / comment removal
const minifyCode = (code: string) => {
  code = removeComments(code);
  code = renameLocals(code);
  code = removeWhiteSpace(code);
  return code;
};

export const symbolDictionary = {
  A: 'at' as 'at',
  B: 'bindings' as 'bindings',
  C: 'vec4<f32>' as 'vec4<f32>',
  D: 'vec3<f32>' as 'vec3<f32>',
  E: 'exports' as 'exports',
  F: 'func' as 'func',
  G: 'flags' as 'flags',
  H: 'inferred' as 'inferred',
  I: 'identifiers' as 'identifiers',
  J: 'imported' as 'imported',
  K: 'imports' as 'imports',
  L: 'linkable' as 'linkable',
  M: 'members' as 'members',
  N: 'name' as 'name',
  O: 'modules' as 'modules',
  P: 'parameters' as 'parameters',
  Q: 'qual' as 'qual',
  R: 'symbol' as 'symbol',
  S: 'symbols' as 'symbols',
  T: 'type' as 'type',
  U: 'struct' as 'struct',
  V: 'variable' as 'variable',
  W: 'visibles' as 'visibles',
  X: 'externals' as 'externals',
  Y: 'types' as 'types',
  Z: 'attr' as 'attr',
};

/** ES/CommonJS Transpiler */
export const transpileWGSL = makeTranspile('wgsl', 'wgsl', symbolDictionary, loadModule, compressAST, minifyCode);

/** Templated literal syntax:

```tsx
wgsl`...`
``` */
export const wgsl = (literals: TemplateStringsArray, ...tokens: string[]) => {
  const code = zip(literals, tokens).flat();
  return loadModuleWithCache(code.join(''));
};

/** Format `number` as WGSL `f32` */
export const f32 = (x: number) => {
  const s = x.toString();
  return (!s.match(/\./)) ? s + '.0' : s;
};
/** Format `number` as WGSL `u32` */
export const u32 = (x: number) => Math.round(x).toString() + 'u';
/** Format `number` as WGSL `i32` */
export const i32 = (x: number) => Math.round(x).toString();
