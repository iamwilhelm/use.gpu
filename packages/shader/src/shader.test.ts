import GLSLModules from '@use-gpu/glsl';
import { parseGLSL } from './shader';

describe('shader', () => {
  
  const code = GLSLModules['instance/quad/vertex'];
  const parsed = parseGLSL(code);
  
});