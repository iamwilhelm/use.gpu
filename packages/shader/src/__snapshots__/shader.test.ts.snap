// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`shader extracts quad fragment declarations 1`] = `
Array [
  Object {
    "node": "[Declaration]",
    "symbols": Array [
      Object {
        "name": "fragColor",
        "node": "[SingleDeclaration]",
      },
    ],
    "variable": Object {
      "locals": Array [
        Object {
          "expr": null,
          "name": "fragColor",
          "node": "[SingleDeclaration]",
        },
      ],
      "node": "[VariableDeclaration]",
      "type": Object {
        "name": "vec4",
        "node": "[QualifiedType]",
        "qualifiers": Array [
          "layout(location = 0)",
          "in",
        ],
      },
    },
  },
  Object {
    "node": "[Declaration]",
    "symbols": Array [
      Object {
        "name": "fragUV",
        "node": "[SingleDeclaration]",
      },
    ],
    "variable": Object {
      "locals": Array [
        Object {
          "expr": null,
          "name": "fragUV",
          "node": "[SingleDeclaration]",
        },
      ],
      "node": "[VariableDeclaration]",
      "type": Object {
        "name": "vec2",
        "node": "[QualifiedType]",
        "qualifiers": Array [
          "layout(location = 1)",
          "in",
        ],
      },
    },
  },
  Object {
    "node": "[Declaration]",
    "symbols": Array [
      Object {
        "name": "outColor",
        "node": "[SingleDeclaration]",
      },
    ],
    "variable": Object {
      "locals": Array [
        Object {
          "expr": null,
          "name": "outColor",
          "node": "[SingleDeclaration]",
        },
      ],
      "node": "[VariableDeclaration]",
      "type": Object {
        "name": "vec4",
        "node": "[QualifiedType]",
        "qualifiers": Array [
          "layout(location = 0)",
          "out",
        ],
      },
    },
  },
]
`;

exports[`shader extracts quad vertex declarations 1`] = `
Array [
  Object {
    "node": "[Declaration]",
    "symbols": Array [
      Object {
        "name": "fragColor",
        "node": "[SingleDeclaration]",
      },
    ],
    "variable": Object {
      "locals": Array [
        Object {
          "expr": null,
          "name": "fragColor",
          "node": "[SingleDeclaration]",
        },
      ],
      "node": "[VariableDeclaration]",
      "type": Object {
        "name": "vec4",
        "node": "[QualifiedType]",
        "qualifiers": Array [
          "layout(location = 0)",
          "out",
        ],
      },
    },
  },
  Object {
    "node": "[Declaration]",
    "symbols": Array [
      Object {
        "name": "fragUV",
        "node": "[SingleDeclaration]",
      },
    ],
    "variable": Object {
      "locals": Array [
        Object {
          "expr": null,
          "name": "fragUV",
          "node": "[SingleDeclaration]",
        },
      ],
      "node": "[VariableDeclaration]",
      "type": Object {
        "name": "vec2",
        "node": "[QualifiedType]",
        "qualifiers": Array [
          "layout(location = 1)",
          "out",
        ],
      },
    },
  },
]
`;

exports[`shader extracts quad vertex functions 1`] = `
Array [
  Object {
    "node": "[FunctionDefinition]",
    "prototype": Object {
      "name": "main",
      "node": "[FunctionPrototype]",
      "parameters": Array [],
      "type": Object {
        "name": "void",
        "node": "[QualifiedType]",
        "qualifiers": Array [],
      },
    },
    "symbols": Array [
      Object {
        "name": "main",
        "node": "[FunctionDefinition]",
      },
    ],
  },
]
`;

exports[`shader extracts quad vertex imports 1`] = `
Array [
  Object {
    "imports": Array [
      Object {
        "imported": "MeshVertex",
        "name": "MeshVertex",
        "node": "[ImportDeclaration]",
      },
    ],
    "name": "use/types",
    "symbols": Array [
      Object {
        "name": "MeshVertex",
        "node": "[ImportDeclaration]",
      },
    ],
  },
  Object {
    "imports": Array [
      Object {
        "imported": "viewUniforms",
        "name": "viewUniforms",
        "node": "[ImportDeclaration]",
      },
      Object {
        "imported": "worldToClip",
        "name": "worldToClip",
        "node": "[ImportDeclaration]",
      },
    ],
    "name": "use/view",
    "symbols": Array [
      Object {
        "name": "viewUniforms",
        "node": "[ImportDeclaration]",
      },
      Object {
        "name": "worldToClip",
        "node": "[ImportDeclaration]",
      },
    ],
  },
  Object {
    "imports": Array [
      Object {
        "imported": "getQuadUV",
        "name": "getQuadUV",
        "node": "[ImportDeclaration]",
      },
    ],
    "name": "geometry/quad",
    "symbols": Array [
      Object {
        "name": "getQuadUV",
        "node": "[ImportDeclaration]",
      },
    ],
  },
]
`;

exports[`shader extracts test declarations 1`] = `
Array [
  Object {
    "node": "[Declaration]",
    "symbols": Array [
      Object {
        "name": "x",
        "node": "[SingleDeclaration]",
      },
      Object {
        "name": "y",
        "node": "[SingleDeclaration]",
      },
    ],
    "variable": Object {
      "locals": Array [
        Object {
          "expr": null,
          "name": "x",
          "node": "[SingleDeclaration]",
        },
        Object {
          "expr": null,
          "name": "y",
          "node": "[SingleDeclaration]",
        },
      ],
      "node": "[VariableDeclaration]",
      "type": Object {
        "name": "float",
        "node": "[QualifiedType]",
        "qualifiers": Array [],
      },
    },
  },
  Object {
    "node": "[Declaration]",
    "symbols": Array [
      Object {
        "name": "a",
        "node": "[SingleDeclaration]",
      },
      Object {
        "name": "b",
        "node": "[SingleDeclaration]",
      },
      Object {
        "name": "c",
        "node": "[SingleDeclaration]",
      },
      Object {
        "name": "d",
        "node": "[SingleDeclaration]",
      },
    ],
    "variable": Object {
      "locals": Array [
        Object {
          "expr": "3",
          "name": "a",
          "node": "[SingleDeclaration]",
        },
        Object {
          "expr": "1",
          "name": "b",
          "node": "[SingleDeclaration]",
        },
        Object {
          "expr": null,
          "name": "c",
          "node": "[SingleDeclaration]",
        },
        Object {
          "expr": null,
          "name": "d",
          "node": "[SingleDeclaration]",
        },
      ],
      "node": "[VariableDeclaration]",
      "type": Object {
        "name": "int",
        "node": "[QualifiedType]",
        "qualifiers": Array [],
      },
    },
  },
]
`;

exports[`shader extracts use view declarations 1`] = `
Array [
  Object {
    "node": "[Declaration]",
    "struct": Object {
      "name": "viewUniforms",
      "node": "[QualifiedStructDeclaration]",
      "struct": Object {
        "members": Array [
          Object {
            "name": "projectionMatrix",
            "node": "[StructDeclaration]",
            "type": Object {
              "name": "mat4",
              "node": "[mat4]",
              "qualifiers": Array [],
            },
          },
          Object {
            "name": "viewMatrix",
            "node": "[StructDeclaration]",
            "type": Object {
              "name": "mat4",
              "node": "[mat4]",
              "qualifiers": Array [],
            },
          },
          Object {
            "name": "viewPosition",
            "node": "[StructDeclaration]",
            "type": Object {
              "name": "vec4",
              "node": "[vec4]",
              "qualifiers": Array [],
            },
          },
          Object {
            "name": "viewResolution",
            "node": "[StructDeclaration]",
            "type": Object {
              "name": "vec2",
              "node": "[vec2]",
              "qualifiers": Array [],
            },
          },
        ],
        "node": "[StructDeclarationList]",
      },
      "type": Object {
        "name": "ViewUniforms",
        "node": "[QualifiedStructDeclaration]",
        "qualifiers": Array [
          "layout(set = 0, binding = 0)",
          "uniform",
        ],
      },
    },
    "symbols": Array [
      Object {
        "name": "viewUniforms",
        "node": "[Declaration]",
      },
    ],
  },
]
`;
